Phase 1 of refactoring pgo data pipeline

davidwrighton opened the pull request:
Phase 1 of replacing existing infrastructure around handling of pgo data with more flexible schema based approach.

The schema based approach allows the JIT to define the form of data needed for instrumentation.
- The schema associates 4 32bit integers with each data collection point (ILOffset, InstrumentationKind, Count, and Other)
  - Rich meaning is attached to InstrumentationKind, and Count
    - InstrumentationKind defines the size and layout of individual instrumentation data items
    - Count allows a single schema item to be repeated
  - ILOffset and Other are not processed in any specific way by the infrastructure

Changes part of this phase
- PgoManager holds arbitrary amount of pgo data instead of a slab
  - Aware of collectible assemblies
  - Match with pgo data utilizes hash of IL body in addition to IL size information for greater accuracy in match
- JIT no longer uses block count apis, and instead uses schema based apis
  - JIT now explicitly defines the shape of data collected for both basic block and type probes
  - The rest of the system handles that without deep knowledge of what those formats are
- Text file format for pgo data updated
- Existing IBC infrastructure adjusted to speak in terms of schema concept
- Uncompressed and binary encoded implementation of Pgo schema handling
- Update SuperPMI to handle new apis

Future Changes for static Pgo
- Move Pgo type handle histogram processing into JIT
- Extract Pgo data from process using Event infrastructure
- Triggers for controlling Pgo data extraction
- Instrumented Pgo processing as part of dotnet-pgo tool
- Pgo data flow in crossgen2

davidwrighton committed:
VM builds

davidwrighton committed:
Zapper updated

davidwrighton committed:
It all builds

davidwrighton committed:
Add InstrumentationData event to the runtime

Improve stable hash to match the other R2R hashes

Use method IL body hash to replace token in pgo data

Reworked data storage for pgo to allow for more efficient lookup

Checkin before rebase

Jit no longer uses block count apis

davidwrighton committed:
Fixup superpmi and remove not yet ready recordPgoInstrumentationBySchemaForAot api

davidwrighton committed:
Correct handling for reading pgo data

davidwrighton committed:
- Handle count schema items correctly
- Handle large integers correctly

davidwrighton committed:
Fix Linux build

davidwrighton committed:
Fix gcc build failure

davidwrighton committed:
Apply formatting patch

AndyAyersMS commented on src/coreclr/vm/pgo.cpp:
```diff
@@ -270,7 +307,7 @@ void PgoManager::ReadPgoData()
     }
 
-    char     buffer[256];
+    char     buffer[16384];
```
Is there any symbolic constant we can use here?

davidwrighton committed:
Initialize m_pgoManager as needed

AndyAyersMS commented on src/coreclr/vm/pgo.cpp:
```diff
@@ -285,206 +322,516 @@ void PgoManager::ReadPgoData()
+        HeaderList *currentHeaderList = m_pgoHeaders;
+        if (currentHeaderList != NULL)
+        {
+            if (!ComparePgoSchemaEquals(currentHeaderList->header.GetData(), currentHeaderList->header.countsOffset, pSchema, countSchemaItems))
```
Does this need to tie into the IL versioning scheme? I don't think the jit knows which IL version is active for a jit request, but presumably the jit interface does (or could know).

davidwrighton replied:
I wouldn't expect so. The pgo instrumentation data structures will be the same if the schema's match up, and the JIT already must tolerate data which isn't completely accurate.

davidwrighton replied:
Now, the reading side may be more interesting, but again, that gets into the question of how bad is it to present incorrectly structured data, if that's really bad we could easily add the IL code hash into the schema to allow for the JIT to check it for correctness or something.

AndyAyersMS commented on src/coreclr/vm/pgo.cpp:
```diff
@@ -285,206 +322,516 @@ void PgoManager::ReadPgoData()
-            *pBlockCounts = &s_PgoData[index + 2];
-            *pCount       = header->recordCount - 2;
-            *pNumRuns     = 1;
+            if (!ComparePgoSchemaEquals(existingData->header.GetData(), existingData->header.countsOffset, pSchema, countSchemaItems))
```
Ditto here, do we need to make sure we're looking at the same IL version the jit is compiling?

davidwrighton replied:
I wouldn't expect so. The pgo instrumentation data structures will be the same if the schema's match up, and the JIT already must tolerate data which isn't completely accurate.

AndyAyersMS commented on src/coreclr/vm/pgo.cpp:
```diff
@@ -285,206 +322,516 @@ void PgoManager::ReadPgoData()
-        index += header->recordCount;
-        methodsChecked++;
+        AllocMemTracker loaderHeapAllocation;
+        pHeaderList = (HeaderList*)loaderHeapAllocation.Track(pMD->GetLoaderAllocator()->GetHighFrequencyHeap()->AllocMem(allocationSize));
```
Seems like the schema could go in the low frequency heap but maybe that's more trouble than it's worth.

davidwrighton replied:
Well, the jit interface api allows for doing things like that (which is why the Offset field is a full size_t), but for now, it would cause problems for the superpmi implementation I've come up with, and its not super clear how valuable the high vs low frequency heap actually is.

AndyAyersMS reviewed (commented):
Overall this looks really good.

Left a few notes.

davidwrighton commented on src/coreclr/vm/pgo.h:
```diff
@@ -3,6 +3,428 @@
+#include "shash.h"
+
+
+enum class PgoInstrumentationKind
```
My original thought was that this one would only have the minimal number of enum values describing marshalling so that it was clear that the JIT defined the meanings of stuff, but honestly, that's not a major benefit. I'll merge the two.

davidwrighton commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -5730,8 +5748,18 @@ void Compiler::compCompileFinish()
+            for (UINT32 iSchema = 0; iSchema < fgPgoSchemaCount; iSchema++)
+            {
+                if ((fgPgoSchema[iSchema].InstrumentationKind == ICorJitInfo::PgoInstrumentationKind::BasicBlockIntCount) &&
+                    (fgPgoSchema[iSchema].ILOffset == 0))
```
My suggestion instead would be to make a unique InstrumentationKind to describe that instead of using a magic ILOffset. Details like this are why there are 3 separate descriptive fields associated with each run of data collection details. This may require the parser in the jit to be able to handle both forms in the future, but that's just the reality of supporting a static pgo system.

davidwrighton commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -2304,41 +2304,64 @@ private void expandRawHandleIntrinsic(ref CORINFO_RESOLVED_TOKEN pResolvedToken,
+                return HRESULT.E_NOTIMPL;
+            }
+
+            // Only allocation of PGO data for the current method is supported.
```
That limitation will remain only be for IBC style instrumentation in R2R dlls. The general purpose instrumentation support will only be present in the runtime. If we drop support for IBC style R2R instrumentation as we will likely do, this code will be removed entirely from crossgen2.

davidwrighton committed:
Fixup comments as requested

davidwrighton committed:
Remove unneccessary extra PgoInstrumentationKind enum

davidwrighton committed:
Remove unnecessary struct PgoInstrumentationSchema duplication

davidwrighton committed:
Extract pgo format processing logic to an independent header
- This will be better prepared for the port to managed
- No need to include these complex templates and such in all vm files

davidwrighton committed:
Correct !FEATURE_PGO stubs

AndyAyersMS reviewed (commented):
Thanks for the updates.

The code in zapinfo still looks wrong to me, it won't build a proper schema.

Would be good to run the jit-experimental CI leg, and maybe also a jitstress leg.

davidwrighton committed:
Fix Zap IBC instrumentation path both reading and writing

davidwrighton committed:
Merge branch 'master' into pgo_prototype

davidwrighton commented:
@AndyAyersMS I've fixed the ZapInfo issues, and verified that traditional IBC continues to work. I've also scheduled the jit-experimental and jitstress runs as requested.

AndyAyersMS commented:
In case you haven't yet looked at the experimental failures -- they seem to be related.

davidwrighton committed:
Fix issues identified by jit experimental run

davidwrighton committed:
Needed to be a custom lock now that it actually does something

davidwrighton committed:
SuperPMI fix 2

davidwrighton committed:
Merge branch 'master' of github.com:dotnet/runtime into pgo_prototype

BruceForstall commented:
@davidwrighton You didn't update the JIT-EE GUID with this. Can you please follow-up with a GUID change ASAP?
